lib/widgets/message_list.dart

```diff
@@ -61,8 +61,10 @@ class _MessageListPageState extends State<MessageListPage> {
               // MediaQuery.removePadding with `removeTop: true`.
               context: context,
 
-              // The compose box pads the bottom inset.
-              removeBottom: true,
+              // The compose box, when present, pads the bottom inset.
+              // TODO this copies the details of when the compose box is shown;
+              //   if those details get complicated, refactor to avoid copying.
+              removeBottom: widget.narrow != const AllMessagesNarrow(),
```

Comment:
```suggestion
              removeBottom: widget.narrow is! AllMessagesNarrow,
```

Because we override `==`, the two forms are actually exactly equivalent. But the `is` form is what appears in the `ComposeBox` logic, and I think is also the right conceptual framing of it. (For example if we somehow had two variants of `AllMessagesNarrow` — perhaps a boolean for whether to exclude muted? — then they'd still presumably both skip the compose box.)

Reply:
Makes sense.

Reply:
> (For example if we somehow had two variants of `AllMessagesNarrow` — perhaps a boolean for whether to exclude muted? — then they'd still presumably both skip the compose box.)

Yeah. I'm a little wary that `AllMessagesNarrow` might not always mean what we think it means. Here's its dartdoc:

```dart
/// The narrow called "All messages" in the UI.
///
/// This does not literally mean all messages, or even all messages
/// that the user has access to: in particular it excludes muted streams
/// and topics.
```

And yet:

1. Doesn't the "All messages" view include even muted messages?
2. When we `.apiEncode()` this narrow and send it to the server for mark-all-as-read, it's taken to mean all messages, including muted ones.

Reply:
> 1. Doesn't the "All messages" message list include even muted messages?

It does, but that's because we haven't implemented muting yet. When we implement #346, one piece of it is to have the muting control
> which messages are shown in stream narrows and the all-messages narrow.



> 2\. When we `.apiEncode()` this narrow and send it to the server for mark-all-as-read, it's taken to mean all messages, including muted ones.

Hmm true. And that's intended behavior: it's the same as the specialized endpoint https://zulip.com/api/mark-all-as-read does, and the same as the web UI does.

So possibly the better way to conceptualize this narrow (vs. what the quoted doc comment says) is that the *narrow* does include muted messages, but the message list only shows a subset of the messages in the narrow, namely those that aren't muted.

I'll think more about that when I take up #346.

Reply:
> It does, but that's because we haven't implemented muting yet.

Huh, TIL (or today I reminded myself) that zulip-mobile's "All messages" view also seems to filter out messages in muted conversations. 😅 (I almost never use that view myself.)